File: backend/app/models/itinerary.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, List, Optional

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from .user import User


class Itinerary(Base):
    __tablename__ = "itineraries"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )

    destination: Mapped[str] = mapped_column(String(255), nullable=False)
    duration_days: Mapped[int] = mapped_column(Integer, nullable=False)
    budget: Mapped[float] = mapped_column(Float, nullable=False)
    travel_style: Mapped[str] = mapped_column(String(20), nullable=False)  # budget, comfort, luxury
    interests: Mapped[List[str]] = mapped_column(ARRAY(String), nullable=False, default=[])
    group_type: Mapped[str] = mapped_column(String(20), nullable=False)  # solo, couple, family, friends

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    # Relationships
    user: Mapped["User"] = relationship("User")
    activities: Mapped[List["ItineraryActivity"]] = relationship(
        "ItineraryActivity", back_populates="itinerary", cascade="all, delete-orphan",
        order_by="ItineraryActivity.day_number, ItineraryActivity.start_time"
    )


class ItineraryActivity(Base):
    __tablename__ = "itinerary_activities"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    itinerary_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("itineraries.id", ondelete="CASCADE"), nullable=False, index=True
    )

    day_number: Mapped[int] = mapped_column(Integer, nullable=False)
    start_time: Mapped[str] = mapped_column(String(5), nullable=False)  # "09:00"
    end_time: Mapped[str] = mapped_column(String(5), nullable=False)    # "11:00"
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    estimated_cost: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    location: Mapped[str] = mapped_column(String(500), nullable=False)
    category: Mapped[str] = mapped_column(String(30), nullable=False)  # food, sightseeing, transport, rest, activity

    community_entry_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("community_entries.id"), nullable=True
    )

    # Relationships
    itinerary: Mapped["Itinerary"] = relationship("Itinerary", back_populates="activities")

File: frontend/src/types/itinerary.ts
export type TravelStyle = "budget" | "comfort" | "luxury";
export type GroupType = "solo" | "couple" | "family" | "friends";

export interface ItineraryActivity {
  id: string;
  day_number: number;
  start_time: string;
  end_time: string;
  title: string;
  description: string;
  estimated_cost: number;
  location: string;
  category: string;
  community_entry_id: string | null;
}

export interface Itinerary {
  id: string;
  user_id: string;
  destination: string;
  duration_days: number;
  budget: number;
  travel_style: TravelStyle;
  interests: string[];
  group_type: GroupType;
  created_at: string;
  updated_at: string;
  activities: ItineraryActivity[];
}

export interface ItineraryListItem {
  id: string;
  destination: string;
  duration_days: number;
  budget: number;
  travel_style: TravelStyle;
  group_type: GroupType;
  created_at: string;
  activity_count: number;
}

export interface GenerateItineraryPayload {
  destination: string;
  duration_days: number;
  budget: number;
  travel_style: TravelStyle;
  interests: string[];
  group_type: GroupType;
}
